lib/LANraragi/Model/Tankoubon.pm

```diff
@@ -0,0 +1,270 @@
+    my $logger = get_logger( "Tankoubon", "lanraragi" );
+
+    # Tankoubons are represented by RG_[timestamp] in DB. Can't wait for 2038!
+    my @tanks = $redis->keys('TANK_??????????');  
+
+    # Jam tanks into an array of hashes
+    my @result;
+    foreach my $key (sort @tanks) {
+        my %data = get_tankoubon($key);
+        push( @result, \%data );
+    }
+
+    # # Only get the first X keys
+    my $keysperpage = LANraragi::Model::Config->get_pagesize;
+
+    # Return total keys and the filtered ones
+    my $total = $#tanks+1;
+    my $start = $page*$keysperpage;
+    my $end = min( $start + $keysperpage - 1, $#result );
+    return ( $total, $#result + 1, @result[ $start .. $end ] );
+
+    #return @result;
+}
+
+# create_tankoubon(name, existing_id)
+#   Create a Tankoubon.
+#   If an existing Tankoubon ID is supplied, said Tankoubon will be updated with the given parameters.
+#   Returns the ID of the created/updated Tankoubon.
+sub create_tankoubon( $name, $tank_id ) {
+
+    my $redis = LANraragi::Model::Config->get_redis;
+
+    # Set all fields of the group object
+    unless ( length($tank_id) ) {
+        $tank_id = "TANK_" . time();
+
+        my $isnewkey = 0;
+        until ($isnewkey) {
+
+            # Check if the group ID exists, move timestamp further if it does
+            if ( $redis->exists($tank_id) ) {
+                $tank_id = "TANK_" . ( time() + 1 );
+            } else {
+                $isnewkey = 1;
+            }
+        }
+    }
+
+    # Default values for new group
+    # Score 0 will be reserved for the name of the tank
+    $redis->zadd( $tank_id, 0, redis_encode($name)); 
+
+    $redis->quit;
+
+    return $tank_id;
+}
+
+# get_tankoubon(tankoubonid, decoded, page)
+#   Returns the Tankoubon matching the given id.
+#   Returns undef if the id doesn't exist.
+sub get_tankoubon($tank_id,$decoded=0,$page=0) {
+
+    my $logger = get_logger( "Tankoubon", "lanraragi" );
+    my $redis  = LANraragi::Model::Config->get_redis;
+    my $keysperpage = LANraragi::Model::Config->get_pagesize;
+
+    if ( $tank_id eq "" ) {
+        $logger->debug("No Tankoubon ID provided.");
+        return ();
+    }
+
+    unless ( length($tank_id) == 15 && $redis->exists($tank_id) ) {
+        $logger->warn("$tank_id doesn't exist in the database!");
+        return ();
+    }
+
+    # Declare some needed variables
+    my %tank;
+    my @archives; 
+    my @limit = split(' ', "LIMIT " . ($keysperpage * $page) . " $keysperpage");
+
+    # Get name
+    my @name = $redis->zrangebyscore($tank_id, 0, 0, qw{LIMIT 0 1});
+    $tank{name} = redis_decode($name[0]);
+
+    # Grab page
+    my %tankoubon = $redis->zrangebyscore($tank_id, 1, "+inf", "WITHSCORES", @limit);
+
+    # Sort and add IDs to archives array
+    foreach my $i (sort { $tankoubon{$a} <=> $tankoubon{$b} } keys %tankoubon) {
+        push( @archives, $i );
+    }
+
+    # Verify if we require decoded files or just IDs
```

Comment:
Having one of the response fields be a completely different data type depending on a parameter is _hella weird_ API design imo -- I think it'd be better here to always include the ID list as `archives`, and have an additional `full_data` field present with the list of archive data if the API was called with the `decoded` parameter.  

I don't think `decoded` is a very good name either as it took me reading up to this implementation to figure out what it did; It's a bit verbose but maybe `include_full_data` would be better?

Reply:
Yes, I know it's weird, but disgracefully I've been in projects where we have to do it so we don't need to make redundants EPs... Anyway, I changed the name of the param to include_full_data, and like you said, it now appends a full_data field with the Archive objects.

Reply:
Ah well, there's nothing disgraceful about being forced to implement bad upper management decisions imo 😤  
Thanks!